module/util/log.go

```diff
@@ -1,25 +1,114 @@
 package util
 
 import (
+	"sync"
+	"sync/atomic"
+	"time"
+
 	"github.com/rs/zerolog"
 )
 
-// LogProgress takes a total and return function such that when called with a 0-based index
-// it prints the progress from 0% to 100% to indicate the index from 0 to (total - 1) has been
-// processed.
-// useful to report the progress of processing the index from 0 to (total - 1)
-func LogProgress(msg string, total int, logger zerolog.Logger) func(currentIndex int) {
-	logThreshold := float64(0)
-	return func(currentIndex int) {
+// LogProgressFunc is a function that can be called to add to the progress
+type LogProgressFunc func(addProgress int)
+
+// LogProgress takes a total and return function such that when called adds the given
+// number to the progress and logs the progress every 10% or every 60 seconds whichever
+// comes first.
+// The returned function can be called concurrently.
+// An eta is also logged, but it assumes that the progress is linear.
+func LogProgress(log zerolog.Logger, msg string, total int) LogProgressFunc {
+	nth := uint32(total / 10) // sample every 10% by default
+	if nth == 0 {
+		nth = 1
+	}
+
+	sampler := log.Sample(newProgressLogsSampler(nth, 60*time.Second))
+
+	start := time.Now()
+	currentIndex := uint64(0)
+	return func(add int) {
+		current := atomic.AddUint64(&currentIndex, uint64(add))
+
 		percentage := float64(100)
 		if total > 0 {
-			percentage = (float64(currentIndex+1) / float64(total)) * 100. // currentIndex+1 assuming zero based indexing
+			percentage = (float64(current) / float64(total)) * 100.
+		}
+		elapsed := time.Since(start)
+		elapsedString := elapsed.Round(1 * time.Second).String()
+
+		etaString := "unknown"
+		if percentage > 0 {
+			eta := time.Duration(float64(elapsed) / percentage * (100 - percentage))
+			if eta < 0 {
+				eta = 0
+			}
+			etaString = eta.Round(1 * time.Second).String()
+
 		}
 
-		// report every 10 percent
-		if percentage >= logThreshold {
-			logger.Info().Msgf("%s progress: %v percent", msg, logThreshold)
-			logThreshold += 10
+		if current != uint64(total) {
+			sampler.Info().Msgf("%s progress %d/%d (%.1f%%) elapsed: %s, eta %s", msg, current, total, percentage, elapsedString, etaString)
```

Comment:
How does this `Info` works? 

It only prints when sampler triggers a sample?

Reply:
Yes. The sampler determines which log line should be printed. Here the sampler logic is every Nth sample (so that a long line is printed every 10%) or the first sample if the time since the last sample is more than a minute.